javascript/javascript2/week1/spirit-animal/main.js

```diff
@@ -0,0 +1,118 @@
+    "chubby",
+    "crazy",
+  ];
+
+  let spiritAnimal = `The ${adjectives[Math.floor(Math.random() * 9)]} ${
+    nouns[Math.floor(Math.random() * 9)]
+  } `;
+  return spiritAnimal;
+}
+
+//Default event (since dropdown.onchange requires some "change" to start working)
+// I tried putting it as a default event in the switch statement but it didn't work - why?
+window.onload = function (event) {
+  button.onclick = function (event) {
+    const name = inputField.value;
+    if (name === "") {
+      outputPar.innerHTML = "Please enter your name";
+      button.innerHTML = "Click me"; //in case one wants to get a new name with an empty string
+    } else {
+      let spiritAnimal = getSpiritAnimal();
+      outputPar.innerHTML = `${name} - ${spiritAnimal}`;
+      button.innerHTML = "Get a new name";
+    }
+  };
+};
+
+//Selector onchange event
+dropdown.onchange = function (event) {
+  const chosenValue = dropdown.value;
+  switch (chosenValue) {
+    case "click":
+      inputField.oninput = null; //removing the oninput event in case of switching between the options
+      inputField.onmouseover = null; //removing the onmouseover event in case of switching between the options
+      button.onclick = function (event) {
+        const name = inputField.value;
+        if (name === "") {
+          outputPar.innerHTML = "Please enter your name";
+          button.innerHTML = "Click me"; //in case one wants to get a new name with an empty string
+        } else {
+          let spiritAnimal = getSpiritAnimal();
+          outputPar.innerHTML = `${name} - ${spiritAnimal}`;
+          button.innerHTML = "Get a new name";
+        }
+      };
+      break;
+
+    case "hover":
+      inputField.oninput = null; // removing the oninput event in case of switching between the options
+      button.onclick = null; // removing the onclick event in case of switching between the options
+      inputField.onmouseover = function (event) {
+        const name = inputField.value;
+        if (name === "") {
+          outputPar.innerHTML = "Please enter your name";
+          button.innerHTML = "Click me"; //in case one wants to get a new name with an empty string
+        } else {
+          let spiritAnimal = getSpiritAnimal();
+          outputPar.innerHTML = `${name} - ${spiritAnimal}`;
+        }
+      };
+      break;
+    case "typing":
+      inputField.onmouseover = null; //removing the onmouseover event in case of switching between the options
+      button.onclick = null; //removing the onclick event in case of switching between the options
+      inputField.oninput = function (event) {
+        const name = inputField.value;
+        if (name === "") {
+          outputPar.innerHTML = "Please enter your name";
+          button.innerHTML = "Click me"; //in case one wants to get a new name with an empty string
+        } else {
+          let spiritAnimal = getSpiritAnimal();
+          outputPar.innerHTML = `${name} - ${spiritAnimal}`;
+        }
```

Comment:
This is great work, but I do have to point out that we're seeing a lot of code duplication here. Here's what I'd like to see: 
 
1. Refactor the event listener into its own function, e.g. 
```javascript
  function inputListener(event) {
        const name = inputField.value;
        if (name === "") {
          outputPar.innerHTML = "Please enter your name";
          button.innerHTML = "Click me"; //in case one wants to get a new name with an empty string
        } else {
          let spiritAnimal = getSpiritAnimal();
          outputPar.innerHTML = `${name} - ${spiritAnimal}`;
        }
   }
```
2. Make a separate function which clears all the event listeners, and then call it before the switch statement. Something like: 
```javascript
function clearEventListeners() {
    document.getElementsByTagName("button")[0].onclick = null;
    document.getElementsByTagName("input")[0].oninput = null;
    document.getElementsByTagName("input")[0].onmouseover = null;
}
```

Then inside each case in the switch statement, you just set the appropriate event listener on the right node. That should make the `dropdown.onchange` function quite simple!

Reply:
You are totally right! Gonna fix it now - thought about it, but I got some bugs while coding, so I was focused on something else. Thanks a lot for your feedback!